### PR TITLE
Add `ORM\Query\SelectQuery`.

### PR DESCRIPTION
**Pull request by ADmad**
This makes the namepsace and classname for SELECT query consistent with other query types. `ORM\Query` is retained as alias for easier upgrading.

### CODE REVIEW

**Comment by ADmad** [comment]
I'll finish up the change like updating the argument/return types if the change is approved in principle.

**Review by markstory — APPROVED** [review]
I think this is fine. We're keeping the `query()` method and we can use rector to rename any user-land references.

File: src/ORM/Query/SelectQuery.php [hunk 1]
```diff
@@ -0,0 +1,1775 @@
...
+     * Setting the resultset of a query will make execute() a no-op. Instead
+     * of executing the SQL query and fetching results, the ResultSet provided to this
+     * method will be returned.
+     *
+     * This method is most useful when combined with results stored in a persistent cache.
+     *
+     * @param iterable $results The results this query should return.
+     * @return $this
+     */
+    public function setResult(iterable $results)
```
Comment (othercorey) [inline src/ORM/Query/SelectQuery.php:213]: Hopefully we can get rid of this.

**Review by othercorey — APPROVED** [review]

File: src/ORM/Table.php [hunk 2]
```diff
@@ -1666,9 +1667,19 @@ protected function _getFindOrCreateQuery(Query|callable|array $search): Query
     /**
      * Creates a new Query instance for a table.
```
Comment (othercorey) [inline src/ORM/Table.php]:
```suggestion
     * Creates a new SelectQuery instance for a table.
```

**Comment by othercorey** [comment]
> I think this is fine. We're keeping the `query()` method and we can use rector to rename any user-land references.

I think this is necessary.

File: src/ORM/Query/SelectQuery.php (hunk 1 above)
Comment (ADmad) [reply src/ORM/Query/SelectQuery.php:213]: I can look into it later, but based on the method's description I am not too hopeful.

**Review by markstory — APPROVED** [review]
Looks good to me. I was trying to think of a way we could make these new classes available in 4.5 so that users could incrementally update before jumping to 5.x. The solutions I came up with would require backporting a significant amount of the new query interface. I could look into how hard that would be to do.

**Comment by othercorey** [comment]
> Looks good to me. I was trying to think of a way we could make these new classes available in 4.5 so that users could incrementally update before jumping to 5.x. The solutions I came up with would require backporting a significant amount of the new query interface. I could look into how hard that would be to do.

We could introduce the Select, Insert, Update and Delete queries that extend ORM\Query and assert they are the correct type when compiling. That would let them work with ORM\Query typed functions, but would mean that part breaks when moving to 5.0.

**Comment by markstory** [comment]
> We could introduce the Select, Insert, Update and Delete queries that extend ORM\Query and assert they are the correct type when compiling. That would let them work with ORM\Query typed functions, but would mean that part breaks when moving to 5.0.

I think that's good enough. The static checkers will ensure that at least `Query` is being used, and we'll have runtime checking for all the other methods. With limited resources for backporting I think that covers the primary scenario we need to worry about. Users can use the runtime checks and their tests to ensure they are forwards compatible. They won't get perfect IDE help but it will at least be the same as what they've had for a long time.

**Pull request merged**